### PR TITLE
Session affinity for network balancer

### PR DESCRIPTION
**Pull request by aleksandrov**
Adds support of SessionAffinity for Google network load balancer.

Related PR for Deck is https://github.com/spinnaker/deck/pull/7070

Original issue is https://github.com/spinnaker/spinnaker/issues/3521

@jtk54 can you please look to both PRs?

### CODE REVIEW

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy [hunk 1]
```diff
@@ -139,6 +142,8 @@ class UpsertGoogleLoadBalancerAtomicOperation extends GoogleAtomicOperation<Map>
              || description.healthCheck && !existingTargetPool.healthChecks
              || !description.healthCheck && existingTargetPool.healthChecks)
 
+        needToUpdateSessionAffinity = description.sessionAffinity != GoogleSessionAffinity.valueOf(existingTargetPool.getSessionAffinity());
```
Comment (jtk54) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy]: Is `description.sessionAffinity` null when when there is no affinity selected? What does the `valueOf()` report from the existing target pool has no session affinity configured?

**Comment by maggieneterval** [comment]
Hey @aleksandrov are you waiting on any further input here? Deck PR looks good, just waiting on the resolution of open comments on this PR to merge both.

**Comment by maggieneterval** [comment]
Hey @aleksandrov just a friendly ping, are you no longer working on this PR?

**Comment by aleksandrov** [comment]
Hi  @maggieneterval it was busy month, but I'm going to spend some time on PR this Friday. Basically I need to look/address @jtk54 comments.

**Comment by maggieneterval** [comment]
@aleksandrov sounds good! :+1:

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/description/UpsertGoogleLoadBalancerDescription.groovy [hunk 2]
```diff
@@ -51,6 +52,9 @@ class UpsertGoogleLoadBalancerDescription extends AbstractGoogleCredentialsDescr
    */
   List<GoogleBackendService> backendServiceDiff
 
+  // NLB attribues.
+  GoogleSessionAffinity sessionAffinity
```
[the remark this answers is not in the document]
Comment (aleksandrov) [reply clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/description/UpsertGoogleLoadBalancerDescription.groovy]: There were two places where `sessionAffinity` field was added:

Write part - `UpsertGoogleLoadBalancerDescription` is used in `UpsertGoogleLoadBalancerAtomicOperation`.
Read part - `GoogleLoadBalancerView` is used in `GoogleNetworkLoadBalancer`.

Please let me know if I missed something.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy (hunk 1 above)
Comment (aleksandrov) [reply clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy]: `description.sessionAffinity` is `NONE` when no affinity selected.

`existingTargetPool.getSessionAffinity()` returns `"NONE"`, `"CLIENT_IP"` or `"CLIENT_IP_PROTO"`. `valueof()` returns corresponding enum value. 

We are comparing enum values here.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy [hunk 3]
```diff
@@ -105,11 +106,13 @@ class UpsertGoogleLoadBalancerAtomicOperation extends GoogleAtomicOperation<Map>
     boolean needToUpdateForwardingRule = false
     boolean needToUpdateTargetPool = false
     boolean needToUpdateHttpHealthCheck = false
+    boolean needToUpdateSessionAffinity = false
```
[the remark this answers is not in the document]
Comment (aleksandrov) [reply clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy]: going to rewrite this today to be a bit more clear

**Comment by maggieneterval** [comment]
@aleksandrov Thank you so much for addressing Jake's comments -- is this ready to be merged?

**Comment by aleksandrov** [comment]
@maggieneterval Not fully yet. Adding more tests now. I think I also need final approval from @jtk54

**Comment by maggieneterval** [comment]
Unfortuantely @jtk54 is no longer with the Spinnaker team :cry: but is there anyone else who usually reviews your PRs that we should tag for a final look?

**Comment by aleksandrov** [comment]
sad news! I've looked to original [issue](https://github.com/spinnaker/spinnaker/issues/3521). There are @fjallot and @lwander in the comments, may be they can give a feedback?

**Comment by maggieneterval** [comment]
@lwander has actually also moved on from the Spinnaker team :sob: but yes, if @fjallot or maybe @seanpeters86 who is also very knowledgeable about the GCE provider want to take a look that would be great!

**Comment by aleksandrov** [comment]
Hey guys what is going on? :-D It is not so scarry PR!

ps. Nevernind, I'm joking :-)

**Comment by maggieneterval** [comment]
Haha it looks good to me and I am happy to merge whenever you are ready @aleksandrov, just wanted to double-check if there was anyone else you'd like to take a look as well :smile:

**Comment by robzienert** [comment]
Ping - what's the status of this PR?

**Comment by maggieneterval** [comment]
@aleksandrov I believe you were going to add a few more tests per @seanpeters86's comment, and then this would be ready to merge?

**Comment by aleksandrov** [comment]
Hi all,

Sorry about slow progress on PR, I plan to have some dedicated time at the
end of August to finish tests.

пн, 12 серп. 2019 р., 17:33 користувач Maggie Neterval <
notifications@github.com> пише:

**Comment by maggieneterval** [comment]
Hey @aleksandrov I see you added more test coverage to these changes, thanks a bunch! Is this ready for another pass before being merged or are more changes coming?

**Comment by aleksandrov** [comment]
Hi @maggieneterval These ones for applying SessionAffinity values, need add atleast one more for reading Session Affinity back. Going to contribute this Friday.

**Comment by aleksandrov** [comment]
Hi @seanpeters86 Can you review once again? I've added tests and I think all things are in place now.

**Comment by maggieneterval** [comment]
@aleksandrov Thanks for pushing this across the finish line! :tada: 

I'm also going to add my amazing Java wizard colleague @plumpy for a final look as well.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy [hunk 4]
```diff
@@ -131,6 +132,19 @@ class UpsertGoogleLoadBalancerAtomicOperation extends GoogleAtomicOperation<Map>
         project, region, GCEUtil.getLocalName(existingForwardingRule.target), compute, task, BASE_PHASE, this)
 
       if (existingTargetPool) {
+        GoogleSessionAffinity newSessionAffinity
+        if (description.sessionAffinity == null) {
```
Comment (plumpy) [inline clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/loadbalancer/UpsertGoogleLoadBalancerAtomicOperation.groovy]: The Java8+ idiom for writing this is

```
GoogleSessionAffinity newSessionAffinity =
    Optional.ofNullable(description.sessionAffinity).orElse(GoogleSessionAffinity.NONE);
```

**Pull request merged**